### PR TITLE
Encapsulate Batch

### PR DESCRIPTION
**Pull request by LordDarkula**
Right now many places in the query execution process access the underlying `DataFrame` in a `Batch`. This tightly couples the code to `Batch` and the pandas library. This code decouples the two and builds a larger API in `Batch` to allow indirect access to `Batch` functionality.

### CODE REVIEW

File: eva/models/storage/batch.py [hunk 1]
```diff
@@ -175,6 +218,29 @@ def sort_orderby(self, by, sort_type=None) -> None:
...
+        self._frames = ~self._frames
+
+    def all_true(self) -> bool:
+        return self._frames.all().bool()
+
+    def all_false(self) -> bool:
+        inverted = ~self._frames
+        return inverted.all().bool()
+
+    def create_mask(self) -> List:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:231]: What is this?

File: eva/models/storage/batch.py [hunk 2]
```diff
@@ -191,6 +257,9 @@ def project(self, cols: None) -> Batch:
             )
         return Batch(self._frames[verfied_cols], self._identifier_column)
 
+    def repeat(self, times: int) -> None:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:260]: Why do we need this?

File: eva/models/storage/batch.py [hunk 3]
```diff
@@ -249,18 +318,90 @@ def concat(cls, batch_list: Iterable[Batch], copy=True) -> Batch:
 
         return Batch(frame)
 
+    @classmethod
+    def inner_join(cls, first: Batch, second: Batch) -> Batch:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:322]: Can we make it join and pass `inner` as an argument? We will need `left` join in an upcoming feature.

File: eva/models/storage/batch.py [hunk 4]
```diff
@@ -249,18 +318,90 @@ def concat(cls, batch_list: Iterable[Batch], copy=True) -> Batch:
...
+        if expression == ExpressionType.ARITHMETIC_ADD:
+            return Batch(pd.DataFrame(first._frames + second._frames))
+        elif expression == ExpressionType.ARITHMETIC_SUBTRACT:
+            return Batch(pd.DataFrame(first._frames - second._frames))
+        elif expression == ExpressionType.ARITHMETIC_MULTIPLY:
+            return Batch(pd.DataFrame(first._frames * second._frames))
+        elif expression == ExpressionType.ARITHMETIC_DIVIDE:
+            return Batch(pd.DataFrame(first._frames / second._frames))
+
+    def hash_items(self, indexes) -> None:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:345]: Rename it to something reflecting that hash values are assigned to `index`

File: eva/models/storage/batch.py [hunk 5]
```diff
@@ -249,18 +318,90 @@ def concat(cls, batch_list: Iterable[Batch], copy=True) -> Batch:
...
+
     def empty(self):
         """Checks if the batch is empty
         Returns:
             True if the batch_size == 0
         """
         return len(self) == 0
 
+    @classmethod
+    def compare_contains(cls, batch1: Batch, batch2: Batch) -> None:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:368]: Move it up with other class methods

**Review by gaurav274 — CHANGES_REQUESTED** [review]

**Comment by gaurav274** [comment]
Great job with the rewrite! I'll take another closer look. 
Please also profile the performance of queries with the new changes. Doing it for integration test cases should be sufficient. I want to ensure we don't add unnecessary copy operations. Thanks!

File: eva/models/storage/batch.py (hunk 1 above)
Comment (LordDarkula) [reply eva/models/storage/batch.py:253]: It is used when evaluating logical expressions.

**Comment by xzdandy** [comment]
Very nice work! Thanks for the efforts.

**Review by jarulraj — APPROVED** [review]

**Pull request merged**